src/Akka.Persistence.Redis/Query/Stages/AllEventsSource.cs

```diff
@@ -0,0 +1,343 @@
+                            switch (_state)
+                            {
+                                case State.NotifiedWhenQuerying:
+                                    // maybe we missed some new event when querying, retry
+                                    _state = State.Idle;
+                                    Query();
+                                    break;
+                                case State.Querying:
+                                    if (_live)
+                                    {
+                                        // nothing new, wait for notification
+                                        _state = State.WaitingForNotification;
+                                    }
+                                    else
+                                    {
+                                        // not a live stream, nothing else currently in the database, close the stream
+                                        CompleteStage();
+                                    }
+                                    break;
+                                default:
+                                    Log.Error($"Unexpected source state: {_state}");
+                                    FailStage(new IllegalStateException($"Unexpected source state: {_state}"));
+                                    break;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        var evts = events.ZipWithIndex().Select(c =>
+                        {
+                            var repr = c.Key.Item2;
+                            if (repr != null && !repr.IsDeleted)
+                            {
+                                return new EventEnvelope(new Sequence(_currentOffset + c.Value), repr.PersistenceId, repr.SequenceNr, repr.Payload);
+                            }
+
+                            return null;
+                        }).ToList();
+
+                        _currentOffset += nb;
+                        if (evts.Count > 0)
+                        {
+                            evts.ForEach(_buffer.Enqueue);
+                            Deliver();
+                        }
+                        else
+                        {
+                            // requery immediately
+                            _state = State.Idle;
+                            Query();
+                        }
+                    }
+                });
+
+                if (_live)
+                {
+                    // subscribe to notification stream only if live stream was required
+                    var messageCallback = GetAsyncCallback<(RedisChannel channel, string bs)>(data =>
+                    {
+                        if (data.channel.Equals(_journalHelper.GetEventsChannel()))
+                        {
+                            Log.Debug("Message received");
+
+                            switch (_state)
```

Comment:
What is the default behaviour for this switch?

Reply:
`GetAsyncCallback` is called here
````csharp
 _subscription.Subscribe(_journalHelper.GetEventsChannel(), (channel, value) =>
                    {
                        messageCallback.Invoke((channel, value));
                    });
````
It is likely that the `_state` is changed when a new message is recieved

Reply:
Add a default behaviour anyway. If you are sure that code execution would never reach default, then have it throw an exception because it is unexpected. It helps to catch unexpected behaviour, instead of having a really hard to debug bug later on.

Reply:
I think to understand you better, I would a code snippet about what you are talking about. Thanks

Reply:
The switch statement doesn't have a `default` label, it is a good idea to add a `default` switch handler just in case something wrong happened, so we can catch it, instead of letting it fell through and causing a bug

Reply:
Oh!!